File: public/cpp/087-merge-k-sorted-lists.cpp
/*
 * Problem: Merge k Sorted Lists (LeetCode 23)
 * Link: https://leetcode.com/problems/merge-k-sorted-lists/
 * Difficulty: Hard
 * Category: Heaps and Priority Queues
 *
 * Description:
 * You are given an array of k linked-lists lists, each linked-list is sorted
 * in ascending order. Merge all the linked-lists into one sorted linked-list
 * and return it.
 *
 * Example 1:
 * Input:  lists = [[1,4,5],[1,3,4],[2,6]]
 * Output: [1,1,2,3,4,4,5,6]
 *
 * Example 2:
 * Input:  lists = []   Output: []
 * Example 3:
 * Input:  lists = [[]]  Output: []
 *
 * Constraints:
 * - k == lists.length
 * - 0 <= k <= 10^4
 * - 0 <= lists[i].length <= 500
 * - -10^4 <= lists[i][j] <= 10^4
 * - lists[i] is sorted in ascending order.
 */

#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

// Helper: build list from vector
ListNode* makeList(vector<int>& v) {
    ListNode dummy(0); ListNode* cur = &dummy;
    for (int x : v) { cur->next = new ListNode(x); cur = cur->next; }
    return dummy.next;
}
// Helper: print list
void printList(ListNode* h) {
    cout << "[";
    while (h) { cout << h->val; if (h->next) cout << ","; h = h->next; }
    cout << "]";
}

/*
 * APPROACH 1: MIN HEAP (PRIORITY QUEUE)
 *
 * Push the head of each list into a min-heap.
 * Each time pop the smallest node, append to result, then push its next node.
 *
 * Time:  O(N log k)  N = total nodes, k = number of lists
 * Space: O(k)  heap holds at most k nodes
 */
class Solution1 {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b){ return a->val > b->val; };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> pq(cmp);
        for (auto* h : lists) if (h) pq.push(h);
        ListNode dummy(0); ListNode* cur = &dummy;
        while (!pq.empty()) {
            auto* node = pq.top(); pq.pop();
            cur->next = node; cur = cur->next;
            if (node->next) pq.push(node->next);
        }
        return dummy.next;
    }
};

/*
 * APPROACH 2: DIVIDE AND CONQUER
 *
 * Merge pairs of lists: [0+1, 2+3, ...], then repeat until one list remains.
 * Each merge is O(n), depth is O(log k) → O(N log k) total.
 *
 * Time:  O(N log k)
 * Space: O(log k)  recursion stack
 */
class Solution2 {
    ListNode* mergeTwoLists(ListNode* a, ListNode* b) {
        ListNode dummy(0); ListNode* cur = &dummy;
        while (a && b) {
            if (a->val <= b->val) { cur->next = a; a = a->next; }
            else                  { cur->next = b; b = b->next; }
            cur = cur->next;
        }
        cur->next = a ? a : b;
        return dummy.next;
    }
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        int n = lists.size();
        while (n > 1) {
            for (int i = 0; i < n / 2; i++)
                lists[i] = mergeTwoLists(lists[i], lists[n - 1 - i]);
            n = (n + 1) / 2;
        }
        return lists[0];
    }
};

/*
 * APPROACH 3: COLLECT + SORT
 *
 * Collect all values, sort, rebuild list.
 *
 * Time:  O(N log N)
 * Space: O(N)
 */
class Solution3 {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        vector<int> vals;
        for (auto* h : lists)
            for (auto* p = h; p; p = p->next) vals.push_back(p->val);
        sort(vals.begin(), vals.end());
        ListNode dummy(0); ListNode* cur = &dummy;
        for (int v : vals) { cur->next = new ListNode(v); cur = cur->next; }
        return dummy.next;
    }
};

/*
 * APPROACH 4: SEQUENTIAL MERGE (BRUTE)
 *
 * Merge list[0] with list[1], result with list[2], etc.
 *
 * Time:  O(k * N) — poor for large k
 * Space: O(1)
 */
class Solution4 {
    ListNode* mergeTwoLists(ListNode* a, ListNode* b) {
        ListNode dummy(0); ListNode* cur = &dummy;
        while (a && b) {
            if (a->val <= b->val) { cur->next = a; a = a->next; }
            else                  { cur->next = b; b = b->next; }
            cur = cur->next;
        }
        cur->next = a ? a : b;
        return dummy.next;
    }
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        ListNode* result = lists[0];
        for (int i = 1; i < (int)lists.size(); i++)
            result = mergeTwoLists(result, lists[i]);
        return result;
    }
};

int main() {
    // [[1,4,5],[1,3,4],[2,6]] → [1,1,2,3,4,4,5,6]
    auto run = [](vector<vector<int>> raw, int approach) {
        vector<ListNode*> lists;
        for (auto& v : raw) lists.push_back(makeList(v));
        ListNode* res = nullptr;
        if      (approach==1) { Solution1 s; res=s.mergeKLists(lists); }
        else if (approach==2) { Solution2 s; res=s.mergeKLists(lists); }
        else if (approach==3) { Solution3 s; res=s.mergeKLists(lists); }
        else                  { Solution4 s; res=s.mergeKLists(lists); }
        cout << "Approach " << approach << ": "; printList(res); cout << "\n";
    };
    for (int i=1;i<=4;i++) run({{1,4,5},{1,3,4},{2,6}}, i);
    for (int i=1;i<=4;i++) run({{}}, i);
    return 0;
}

/*
 * COMPLEXITY SUMMARY:
 * Approach 1 (Min Heap):         Time O(N log k)  Space O(k)
 * Approach 2 (Divide & Conquer): Time O(N log k)  Space O(log k)
 * Approach 3 (Collect + Sort):   Time O(N log N)  Space O(N)
 * Approach 4 (Sequential):       Time O(k*N)      Space O(1)
 */

// Made with Bob

File: src/data/tutorials/087-merge-k-sorted-lists.js
/**
 * Tutorial content for #087 — Merge K Sorted Lists
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an array of k sorted linked lists, merge them all into one sorted linked list and return it. Each list is sorted in ascending order. This is the same fundamental problem as "Merge K Sorted Lists" (LeetCode #23) — note that multiple problem numbers may map to the same core algorithm.`,
    example: `Input: [[1→4→5], [1→3→4], [2→6]]\nMerge all three sorted lists into one:\n→ Pick min across heads at each step: 1,1,2,3,4,4,5,6\n✅ Answer: 1→1→2→3→4→4→5→6`,
    keyInsight: `At each step you need the smallest current head across all k lists. A min-heap of size k lets you find the minimum in O(log k) instead of scanning all k heads in O(k), giving O(n log k) total.`,
  },

  approaches: {
    'Min Heap (Priority Queue)': {
      intuition: `Push the head node of each non-null list into a min-heap. Repeatedly extract the minimum node, append it to the result, and if that node has a next node, push the next node into the heap. Repeat until the heap is empty.`,
      steps: [
        `Create a min-heap sorted by node value.`,
        `For each list's head node (if not null): add to heap.`,
        `Create a dummy head node for the result.`,
        `While heap is not empty:`,
        `  node = heap.extractMin()`,
        `  Append node to result list.`,
        `  If node.next exists: heap.add(node.next).`,
        `Return dummy.next.`,
      ],
      example: `Lists: [1→4→5], [1→3→4], [2→6]\n\nHeap initially: {1(list1), 1(list2), 2(list3)}\nExtract 1(list1): result→1; push 4. Heap:{1(L2),2(L3),4(L1)}\nExtract 1(list2): result→1→1; push 3. Heap:{2(L3),3(L2),4(L1)}\nExtract 2(list3): result→1→1→2; push 6. Heap:{3(L2),4(L1),6(L3)}\nExtract 3: push 4. Heap:{4(L1),4(L2),6}\nExtract 4(L1): push 5. ...\nFinal: 1→1→2→3→4→4→5→6 ✅`,
      keyInsight: `O(n log k) time where n = total nodes, k = number of lists. O(k) space for the heap. This is the optimal and standard interview solution for merging k sorted sequences.`,
    },

    'Divide and Conquer': {
      intuition: `Treat the k lists like a tournament bracket. Merge pairs of lists in round 1 (getting k/2 merged lists), then merge those pairs in round 2, and so on — log k rounds total. Each round does O(n) work total. This is the merge sort approach applied to k-way merging.`,
      steps: [
        `If 0 or 1 list: return as-is.`,
        `Pair up adjacent lists and merge each pair using the standard two-list merge.`,
        `Replace the list array with the merged results (half as many).`,
        `Repeat until only one list remains.`,
      ],
      example: `Round 0: [L1,L2,L3] → merge(L1,L2)=[1,1,3,4,4,5], L3=[2,6]\nRound 1: [L12, L3] → merge L12 and L3\n= 1→1→2→3→4→4→5→6 ✅`,
      keyInsight: `O(n log k) time, O(log k) space for the recursion stack. Same complexity as the heap approach but with better cache performance on some hardware. Often the "elegant" solution interviewers appreciate.`,
    },

    'Collect and Sort': {
      intuition: `The simplest possible approach: collect all values from all lists into a single array, sort the array, then build a new linked list from the sorted values. No heaps, no merging logic — just collect, sort, rebuild.`,
      steps: [
        `Traverse all k lists and collect all node values into an array.`,
        `Sort the array.`,
        `Build a new linked list from the sorted array.`,
        `Return its head.`,
      ],
      example: `Lists: [1→4→5], [1→3→4], [2→6]\nCollect: [1,4,5,1,3,4,2,6]\nSort: [1,1,2,3,4,4,5,6]\nBuild: 1→1→2→3→4→4→5→6 ✅`,
      keyInsight: `O(n log n) time, O(n) space. Simple to code but discards the "already sorted" property of each list — we're paying to re-sort information that was partially sorted for free. Fine for quick solutions in interviews.`,
    },

    'Sequential Merge (Brute Force)': {
      intuition: `Merge the lists one by one: merge list1 with list2, take the result and merge with list3, and so on. This is the simplest iterative approach using the standard two-list merge as a subroutine.`,
      steps: [
        `Start with result = lists[0].`,
        `For i from 1 to k-1: result = mergeTwoLists(result, lists[i]).`,
        `Return result.`,
      ],
      example: `Lists: [L1=[1,4,5], L2=[1,3,4], L3=[2,6]]\n\nmerge(L1, L2) = [1,1,3,4,4,5]\nmerge([1,1,3,4,4,5], L3=[2,6]) = [1,1,2,3,4,4,5,6] ✅`,
      keyInsight: `O(kn) time (each of the n total nodes is touched in up to k merges), O(1) extra space. Much slower than the heap/divide-and-conquer for large k. Acceptable only for small k (like k ≤ 10).`,
    },
  },
}
